### ISSUE TITLE
flowBeginTime may not always be valid

### ISSUE THREAD

**Issue by philbooth** [mozilla/fxa-auth-server#1459]
As pointed out by https://github.com/mozilla/fxa-auth-server/issues/1457#issuecomment-247489268:

> in the linked graph, I notice that the vast majority of the records are on fennec, and some are on very old versions of desktop firefox.  I wonder if this is something to do with placeholder device records not properly emitting the `device.created` event?

The following query shows these records:

https://sql.telemetry.mozilla.org/queries/1228/source

**Comment by philbooth** [comment]
I've now verified that this is a real problem.

**Comment by philbooth** [comment]
> I've now verified that this is a real problem.

Wait, no I didn't, I only verified that I can't read logs correctly.

**Comment by philbooth** [comment]
> > I've now verified that this is a real problem.
>
> Wait, no I didn't, I only verified that I can't read logs correctly.

Hang on, wait again, maybe I *did* read the logs correctly first time.

It looks like this is another weird intermittent thing. The next two log excerpts are from legacy device `/certificate/sign` requests, recorded consecutively in the same auth server process from the same device.

In this one, all of the events are present and correct:

```
auth-server key server-2 fxa-auth-server.INFO: SessionToken.isFresh {"op":"SessionToken.isFresh","uid":{"type":"Buffer","data":[201,132,53,201,178,96,78,254,179,177,208,208,137,34,199,8]},"tokenAge":15186,"fresh":true}
auth-server key server-2 fxa-auth-server.INFO: activityEvent {"event":"device.created","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","uid":"c98435c9b2604efeb3b1d0d08922c708","device_id":"0243dbe21f59e9aa1416316d115c9ae3","is_placeholder":true}
auth-server key server-2 fxa-auth-server.INFO: flowEvent {"event":"device.created","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","time":1475175330117,"flow_id":"2ae001625fe47387bc04a7840fb8004034dedaa364362e77a6a70f53ae6104e4","flow_time":19574,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
auth-server key server-2 fxa-auth-server.INFO: device:createPlaceholder {"op":"device:createPlaceholder","uid":{"type":"Buffer","data":[201,132,53,201,178,96,78,254,179,177,208,208,137,34,199,8]},"id":{"type":"Buffer","data":[2,67,219,226,31,89,233,170,20,22,49,109,17,92,154,227]}}
auth-server key server-2 fxa-auth-server.ERROR: push.pushToDevices {"op":"push.pushToDevices","uid":{"type":"Buffer","data":[201,132,53,201,178,96,78,254,179,177,208,208,137,34,199,8]},"deviceId":"ea7ea1f191ee17ab6feabf8224473ed7","err":"Error: No Push Callback"}
auth-server key server-2 fxa-auth-server.INFO: push.pushToDevices {"op":"push.pushToDevices","name":"push.device_connected.no_push_callback"}
auth-server key server-2 {"event":"device:create","data":{"uid":"c98435c9b2604efeb3b1d0d08922c708","id":"0243dbe21f59e9aa1416316d115c9ae3","timestamp":1475175330114,"isPlaceholder":true,"metricsContext":{}}}
auth-server key server-2 fxa-auth-server.INFO: activityEvent {"event":"account.signed","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","uid":"c98435c9b2604efeb3b1d0d08922c708","account_created_at":1475175314927,"device_id":"0243dbe21f59e9aa1416316d115c9ae3"}
auth-server key server-2 fxa-auth-server.INFO: flowEvent {"event":"account.signed","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","time":1475175330123,"flow_id":"2ae001625fe47387bc04a7840fb8004034dedaa364362e77a6a70f53ae6104e4","flow_time":19580,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
auth-server key server-2 fxa-auth-server.INFO: request.summary {"op":"request.summary","code":200,"errno":0,"rid":"1475175330109:pbooth-22205.home:43138:itop7oug:10032","path":"/v1/certificate/sign","lang":"en-US, en","agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","remoteAddressChain":["127.0.0.1"],"t":15,"uid":"c98435c9b2604efeb3b1d0d08922c708"}
```

In this one, there are no flow events for either `device.created` or `account.signed`:

```
auth-server key server-2 fxa-auth-server.INFO: activityEvent {"event":"device.created","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","uid":"c98435c9b2604efeb3b1d0d08922c708","device_id":"ecbe71daaff9a8da5d034210e2797d6c","is_placeholder":true}
auth-server key server-2 fxa-auth-server.INFO: device:createPlaceholder {"op":"device:createPlaceholder","uid":{"type":"Buffer","data":[201,132,53,201,178,96,78,254,179,177,208,208,137,34,199,8]},"id":{"type":"Buffer","data":[236,190,113,218,175,249,168,218,93,3,66,16,226,121,125,108]}}
auth-server key server-2 {"event":"device:create","data":{"uid":"c98435c9b2604efeb3b1d0d08922c708","id":"ecbe71daaff9a8da5d034210e2797d6c","timestamp":1475175223316,"isPlaceholder":true,"metricsContext":{}}}
auth-server key server-2 fxa-auth-server.INFO: activityEvent {"event":"account.signed","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","uid":"c98435c9b2604efeb3b1d0d08922c708","account_created_at":1475175193056,"device_id":"ecbe71daaff9a8da5d034210e2797d6c"}
auth-server key server-2 fxa-auth-server.ERROR: push.pushToDevices {"op":"push.pushToDevices","uid":{"type":"Buffer","data":[201,132,53,201,178,96,78,254,179,177,208,208,137,34,199,8]},"deviceId":"ea7ea1f191ee17ab6feabf8224473ed7","err":"Error: No Push Callback"}
auth-server key server-2 fxa-auth-server.INFO: push.pushToDevices {"op":"push.pushToDevices","name":"push.device_connected.no_push_callback"}
auth-server key server-2 fxa-auth-server.INFO: request.summary {"op":"request.summary","code":200,"errno":0,"rid":"1475175223311:pbooth-22205.home:43138:itop7oug:10016","path":"/v1/certificate/sign","lang":"en-US, en","agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","remoteAddressChain":["127.0.0.1"],"t":16,"uid":"c98435c9b2604efeb3b1d0d08922c708"}
```

Scanning back further up the log, there are no flow events for `account.keyfetch` or `account.confirmed` either. And then further back still, I get to this:

```
auth-server key server-2 fxa-auth-server.WARN: metrics.context.validate {"op":"metrics.context.validate","valid":false,"reason":"expired flowBeginTime","agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0"}
auth-server key server-2 fxa-auth-server.INFO: activityEvent {"event":"account.login","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","service":"sync","uid":"c98435c9b2604efeb3b1d0d08922c708"}
auth-server key server-2 fxa-auth-server.ERROR: log.flowEvent {"op":"log.flowEvent","event":"account.login","missingFlowId":true}
```

So it looks like the content server has failed to send valid metrics context data for some reason.

**Comment by philbooth** [comment]
I don't know what I did yesterday to make this fail but, whatever it was, I'm not able to do it now. Popping this on to the back-burner so I can do something less heisenbuggy.

**Comment by rfk** [comment]
> "reason":"expired flowBeginTime"

Possible that you left the login page sitting around for too long before submitting it?  I wonder if we should reconsider this "throw away flows if the flowBeginTime was too far in the past" logic.

**Comment by philbooth** [comment]
> Possible that you left the login page sitting around for too long before submitting it?

Unfortunately I don't think so; I was just hammering away at a sign in -> disconnect -> sign-in loop, so none of the pages were open long.

> I wonder if we should reconsider this "throw away flows if the flowBeginTime was too far in the past" logic.

I have a note to ask you about exactly this in our next 1:1, as it happens. It also made me realise that those longer-than-30-minutes-to-account-signed flows are of course completely fine, it is just people who've had the sign-in page open for a while before submitting. The real limit is an hour, 30 minutes to expire the `flowBeginTime` and 30 minutes in memcached, and sure enough there are no recorded events that are older than that.

**Comment by philbooth** [comment]
[Now that `device.created` is not a flow event](https://github.com/mozilla/fxa-auth-server/pull/1483), this issue no longer exists for flow events.

And in my testing, I've seen no evidence that the `device.created` activity event is not emitted for placeholder devices.

So, even though it feels like I'm cheating a bit, this issue seems redundant now. Closing.

**Comment by philbooth** [comment]
> So, even though it feels like I'm cheating a bit, this issue seems redundant now. Closing.

Actually, that totally was cheating. The real issue seen here was the weird `flowBeginTime`, not the missing event. Re-opening and changing the title to reflect that.

**Comment by philbooth** [comment]
This was resolved by mozilla/fxa-content-server#4383.
